packages/hardhat-core/test/internal/hardhat-network/provider/TransactionQueue.ts

```diff
@@ -0,0 +1,303 @@
+        assert.equal(queue.getNextTransaction(), tx4.data);
+        assert.equal(queue.getNextTransaction(), tx3.data);
+        assert.equal(queue.getNextTransaction(), tx2.data);
+        assert.equal(queue.getNextTransaction(), tx1.data);
+      });
+
+      it("Should not include transactions from a sender whose next tx was discarded", function () {
+        const baseFee = new BN(20);
+
+        const senderWithFirstTxNotMined =
+          "0x0000000000000000000000000000000000000001";
+
+        const senderWithSecondTxNotMined =
+          "0x0000000000000000000000000000000000000002";
+
+        // Effective miner fee: 80
+        const tx1 = createTestTransaction({
+          gasPrice: 100,
+        });
+
+        // Effective miner fee: 79
+        const tx2 = createTestTransaction({
+          maxPriorityFeePerGas: 79,
+          maxFeePerGas: 1000,
+        });
+
+        // Effective miner fee: 78
+        const tx3 = createTestTransaction({
+          maxPriorityFeePerGas: 97,
+          maxFeePerGas: 98,
+        });
+
+        // Not mined
+        // Effective miner fee: 77
+        const tx4 = createTestTransaction({
+          gasPrice: 97,
+          from: senderWithFirstTxNotMined,
+        });
+
+        // Discarded
+        // Effective miner fee: 76
+        const tx5 = createTestTransaction({
+          gasPrice: 96,
+          from: senderWithFirstTxNotMined,
+        });
+
+        // Effective miner fee: 75
+        const tx6 = createTestTransaction({
+          gasPrice: 95,
+          from: senderWithSecondTxNotMined,
+        });
+
+        // Not mined
+        // Effective miner fee: 74
+        const tx7 = createTestTransaction({
+          gasPrice: 94,
+          from: senderWithSecondTxNotMined,
+        });
+
+        // Discarded
+        // Effective miner fee: 73
+        const tx8 = createTestTransaction({
+          gasPrice: 93,
+          from: senderWithSecondTxNotMined,
+        });
+
+        // Discarded
+        // Effective miner fee: 72
+        const tx9 = createTestTransaction({
+          maxFeePerGas: 92,
+          maxPriorityFeePerGas: 80,
+          from: senderWithSecondTxNotMined,
+        });
+
+        // Effective miner fee: 71
+        const tx10 = createTestTransaction({
+          gasPrice: 91,
+        });
+
+        const queue = new TransactionQueue(
+          makeOrderedTxMap([tx1, tx2, tx3, tx4, tx5, tx6, tx7, tx8, tx9, tx10]),
+          baseFee
+        );
+
+        assert.equal(queue.getNextTransaction(), tx1.data);
+        assert.equal(queue.getNextTransaction(), tx2.data);
+        assert.equal(queue.getNextTransaction(), tx3.data);
+        assert.equal(queue.getNextTransaction(), tx4.data);
+        queue.removeLastSenderTransactions();
+
+        assert.equal(queue.getNextTransaction(), tx6.data);
+        assert.equal(queue.getNextTransaction(), tx7.data);
+        assert.equal(queue.getNextTransaction(), tx8.data);
+        queue.removeLastSenderTransactions();
```

Comment:
This doesn't match the description in the comments, I think you meant to call this after tx7?

Reply:
Right. Good catch!